ScaleResiliencyTest$ScaleDown.shouldContinueScaleDownAfterRestart is flaky

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? 
  - Once so far.
- Does it block your work? 
  - Not really.
- Do we suspect that it is a real failure? 
  - Not sure.

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
Assertion condition [List check single element] 
Expecting actual:
  class BrokerState {
    id: 2
    state: ACTIVE
    version: 2
    lastUpdatedAt: 2023-11-16T09:13:32.135700584Z
    partitions: []
}
to match 'Broker 2 has partition 2' predicate within 10 seconds.
</pre>
</details>

**Hypotheses**

> List any hypotheses if you have one; can be ommitted
/

**Logs**

> If possible, provide more context here, e.g. standard output logs, link to build, etc.

<details><summary>Logs</summary>
<pre>
Error:  Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 28.37 s <<< FAILURE! -- in io.camunda.zeebe.it.clustering.dynamic.ScaleResiliencyTest$ScaleDown
Error:  io.camunda.zeebe.it.clustering.dynamic.ScaleResiliencyTest$ScaleDown.shouldContinueScaleDownAfterRestart -- Time elapsed: 28.37 s <<< ERROR!
org.awaitility.core.ConditionTimeoutException: 
Assertion condition [List check single element] 
Expecting actual:
  class BrokerState {
    id: 2
    state: ACTIVE
    version: 2
    lastUpdatedAt: 2023-11-16T09:13:32.135700584Z
    partitions: []
}
to match 'Broker 2 has partition 2' predicate within 10 seconds.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:167)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
	at io.camunda.zeebe.it.clustering.dynamic.ScaleResiliencyTest$ScaleDown.shouldContinueScaleDownAfterRestart(ScaleResiliencyTest.java:210)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:204)
	at org.awaitility.core.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:101)
	at org.awaitility.core.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:81)
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:103)
	... 12 more
</pre>
</details>

[Test result archive](https://drive.google.com/file/d/1iVtdYDtFkpBmLCzvspbXvTVPADR4btf3/view?usp=drive_link) (internal drive link)

--- a/atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java
+++ b/atomix/cluster/src/main/java/io/atomix/raft/cluster/impl/RaftClusterContext.java
@@ -165,8 +165,10 @@ public final class RaftClusterContext implements RaftCluster, AutoCloseable {
               .filter(context -> newMembers.contains(context.getMember()))
               .collect(Collectors.toCollection(ArrayList::new));
 
-      final var oldQuorum = getQuorumFor(oldContexts, calculateMemberValue);
-      final var newQuorum = getQuorumFor(newContexts, calculateMemberValue);
+      final var oldQuorum =
+          getQuorumFor(oldContexts, calculateMemberValue, oldMembers.contains(localMember));
+      final var newQuorum =
+          getQuorumFor(newContexts, calculateMemberValue, newMembers.contains(localMember));
       if (oldQuorum.isPresent() && newQuorum.isPresent()) {
         return Optional.of(Comparators.min(oldQuorum.get(), newQuorum.get()));
       } else if (oldQuorum.isPresent()) {
@@ -176,21 +178,25 @@ public final class RaftClusterContext implements RaftCluster, AutoCloseable {
       }
     }
 
-    return getQuorumFor(contexts, calculateMemberValue);
+    return getQuorumFor(
+        contexts, calculateMemberValue, configuration.newMembers().contains(localMember));
   }
 
   private <T extends Comparable<T>> Optional<T> getQuorumFor(
       final List<RaftMemberContext> contexts,
-      final Function<RaftMemberContext, T> calculateMemberValue) {
+      final Function<RaftMemberContext, T> calculateMemberValue,
+      final boolean includeLocalMemberInQuorum) {
     if (contexts.isEmpty()) {
       return Optional.empty();
     }
     contexts.sort(Comparator.comparing(calculateMemberValue).reversed());
 
     final var remoteActiveMembers = contexts.size();
-    final var totalActiveMembers = remoteActiveMembers + 1;
+    final int includeLocalMember = includeLocalMemberInQuorum ? 1 : 0;
+    final var totalActiveMembers = remoteActiveMembers + includeLocalMember;
     final var quorum = (totalActiveMembers / 2) + 1;
-    final var remoteQuorumIndex = quorum - 1 - 1;
+
+    final var remoteQuorumIndex = quorum - 1 - includeLocalMember;
     final var context = contexts.get(remoteQuorumIndex);
     return Optional.of(calculateMemberValue.apply(context));
   }

--- a/atomix/cluster/src/test/java/io/atomix/raft/ReconfigurationTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/ReconfigurationTest.java
@@ -36,6 +36,7 @@ import java.util.List;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Stream;
 import org.awaitility.Awaitility;
@@ -423,6 +424,34 @@ final class ReconfigurationTest {
       final var expected = List.of(new DefaultRaftMember(id1, Type.ACTIVE, Instant.now()));
 
       assertThat(m1.cluster().getMembers()).containsExactlyInAnyOrderElementsOf(expected);
+    }
+
+    @Test
+    void cannotLeaveWhenNewConfigurationDoesNotHaveQuorum(@TempDir final Path tmp) {
+      // given - a cluster with 2 members
+      final var id1 = MemberId.from("1");
+      final var id2 = MemberId.from("2");
+      final var id3 = MemberId.from("3");
+
+      final var m1 = createServer(tmp, createMembershipService(id1, id2, id3));
+      final var m2 = createServer(tmp, createMembershipService(id2, id1, id3));
+      final var m3 = createServer(tmp, createMembershipService(id3, id1, id2));
+
+      CompletableFuture.allOf(
+              m1.bootstrap(id1, id2, id3), m2.bootstrap(id1, id2, id3), m3.bootstrap(id1, id2, id3))
+          .join();
+
+      // when -  existing member try to leave, when there is no majority for the new configuration
+      m3.shutdown().join();
+      // To reduce chances of flakiness ensure that there is a leader before sending leave
+      awaitLeader(m1, m2);
+
+      // then
+      assertThat(m2.leave())
+          .describedAs(
+              "Should fail to leave because quorum not available for the new configuration")
+          .failsWithin(Duration.ofSeconds(2))
+          .withThrowableOfType(ExecutionException.class);
     }
 
     @Test

--- a/atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java
+++ b/atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java
@@ -260,6 +260,29 @@ final class RaftClusterContextTest {
   }
 
   @Test
+  void shouldNotIncludeLocalMemberInQuorumWhenItIsNotPartOfNewConfiguration() {
+    // given
+    final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());
+    final var remoteMembers =
+        List.<RaftMember>of(
+            new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+            new DefaultRaftMember(new MemberId("3"), Type.ACTIVE, Instant.now()));
+
+    final var raft =
+        raftWithStoredConfiguration(
+            new Configuration(1, 1, Instant.now().toEpochMilli(), remoteMembers));
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+    context.bootstrap(List.of()).join();
+
+    // when
+    context.getMemberContext(new MemberId("2")).setMatchIndex(2);
+    context.getMemberContext(new MemberId("3")).setMatchIndex(3);
+
+    // then
+    assertThat(context.getQuorumFor(RaftMemberContext::getMatchIndex)).hasValue(2L);
+  }
+
+  @Test
   void quorumWhenNewMembersAreAhead() {
     // given
     final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());

